clis/teliod/src/main.rs

```diff
@@ -129,19 +207,193 @@ impl CommandListener {
+    config.hw_identifier = Some(client_config.hw_identifier.clone());
+    config.machine_identifier = Some(client_config.machine_identifier.clone());
+
+    let mut file = File::create(config_path)?;
+    serde_json::to_writer(&mut file, &config)?;
+    Ok(client_config)
+}
+
 // From async context Telio needs to be run in separate task
-fn telio_task() {
-    let _telio = Device::new(
+fn telio_task(
+    client_config: ClientConfig,
+    mut rx_channel: mpsc::Receiver<TelioTaskCmd>,
+    tx_channel: mpsc::Sender<TelioTaskCmd>,
+) -> Result<(), TeliodError> {
+    debug!("Initializing telio device");
+    let mut telio = Device::new(
         Features::default(),
         // TODO: replace this with some real event handling
         move |event| info!("Incoming event: {:?}", event),
         None,
-    );
+    )?;
+
+    let config_ptr = std::sync::Arc::new(client_config);
+    // TODO: This is temporary to be removed later on when we have proper integration
+    // tests with core API. This is to not look for tokens in a test environment
+    // right now as the values are dummy and program will not run as it expects
+    // real tokens.
+    if !config_ptr.auth_token.eq("") {
+        start_telio(&mut telio, config_ptr.private_key)?;
+        get_meshmap(config_ptr.clone(), tx_channel.clone());
+
+        while let Some(cmd) = rx_channel.blocking_recv() {
+            info!("Got command {:?}", cmd);
+            match cmd {
+                TelioTaskCmd::GetMeshmap => get_meshmap(config_ptr.clone(), tx_channel.clone()),
+                TelioTaskCmd::UpdateMeshmap(map) => {
+                    if let Err(e) = telio.set_config(&Some(map)) {
+                        error!("Unable to set meshmap due to {e}");
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+#[allow(mpsc_blocking_send)]
+fn get_meshmap(client_config: Arc<ClientConfig>, tx: mpsc::Sender<TelioTaskCmd>) {
+    let config_clone = client_config.clone();
+    tokio::spawn(async move {
+        let result = get_meshmap_from_server(config_clone).await;
+        match result {
+            Ok(map) => {
+                let meshmap: MeshMap = match serde_json::from_str(&map) {
+                    Ok(map) => map,
+                    Err(e) => {
+                        error!("Unable to parse meshmap due to {e}");
+                        return;
+                    }
+                };
+                trace!("Meshmap {:#?}", meshmap);
+                if let Err(e) = tx.send(TelioTaskCmd::UpdateMeshmap(meshmap)).await {
+                    error!("Unable to send meshmap due to {e}");
+                }
+            }
+            Err(e) => error!("Getting meshmap failed due to {e}"),
+        }
+    });
 }
 
-async fn daemon_event_loop(config: TeliodDaemonConfig) -> Result<(), TeliodError> {
+fn start_telio(telio: &mut Device, private_key: SecretKey) -> Result<(), DeviceError> {
+    if !telio.is_running() {
```

Comment:
Can Telio be running at this point? It might look reasonable locally, but this function is used only once when Telio was just created, so I guess you don't need to check that (I don't think this function is particularly helpful tbh)

Reply:
I would prefer to keep this function, though modified it a bit. A little less clutter in the main function.